File: webapp/backend/analytics/src/main/java/tom/analytics/service/AnalyticsService.java
package tom.analytics.service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import tom.analytics.entity.AssistantDailyUsage;
import tom.analytics.entity.AssistantLeaderboard;
import tom.analytics.entity.AssistantPopularity;
import tom.analytics.entity.ConversationStatistics;
import tom.analytics.entity.DailySystemStatistics;
import tom.analytics.entity.LlmAssistantMetrics;
import tom.analytics.entity.LlmDailyMetrics;
import tom.analytics.entity.LlmHourlyMetrics;
import tom.analytics.entity.LlmSystemOverview;
import tom.analytics.entity.LlmUserMetrics;
import tom.analytics.entity.SystemOverview;
import tom.analytics.entity.UserActionSummary;
import tom.analytics.entity.UserActivitySummary;
import tom.analytics.entity.UserAssistantSummary;
import tom.analytics.entity.UserLoginSummary;
import tom.analytics.entity.UserWorkflowSummary;
import tom.analytics.entity.WorkflowPopularity;
import tom.api.AssistantId;
import tom.api.ConversationId;
import tom.api.UserId;
import tom.api.WorkflowId;

public interface AnalyticsService {
	/*
	 * -------------------------------------------------------------------------
	 * Event recording
	 * -------------------------------------------------------------------------
	 */

	void recordUserLogin(UserId userId);

	void recordUserLogin(UserId userId, LocalDateTime loginTime);

	ConversationStatistics startConversation(UserId userId, ConversationId conversationId);

	ConversationStatistics startConversation(UserId userId, ConversationId conversationId, LocalDateTime started);

	ConversationStatistics recordMessage(UserId userId, ConversationId conversationId);

	ConversationStatistics recordMessage(UserId userId, ConversationId conversationId, LocalDateTime activityTime);

	ConversationStatistics updateConversationActivity(UserId userId, ConversationId conversationId,
			LocalDateTime activityTime);

	ConversationStatistics completeConversation(UserId userId, ConversationId conversationId);

	ConversationStatistics completeConversation(UserId userId, ConversationId conversationId, LocalDateTime completed);

	void recordAssistantCreated(UserId userId, AssistantId assistantId);

	void recordWorkflowCreated(UserId userId, WorkflowId workflowId);

	void recordWorkflowExecuted(UserId userId, WorkflowId workflowId);

	/*
	 * -------------------------------------------------------------------------
	 * User reporting
	 * -------------------------------------------------------------------------
	 */

	UserLoginSummary getUserLoginSummary(UUID userId);

	List<UserLoginSummary> getUserLoginSummaries();

	UserActivitySummary getUserActivity(UUID userId);

	List<UserActivitySummary> getUserActivity();

	List<UserAssistantSummary> getUserAssistantUsage(UUID userId);

	/*
	 * -------------------------------------------------------------------------
	 * Assistant reporting
	 * -------------------------------------------------------------------------
	 */

	AssistantPopularity getAssistantPopularity(UUID assistantId);

	List<AssistantPopularity> getAssistantPopularity();

	List<AssistantDailyUsage> getAssistantDailyUsage(LocalDate from, LocalDate to);

	List<AssistantDailyUsage> getAssistantDailyUsage(UUID assistantId, LocalDate from, LocalDate to);

	List<AssistantLeaderboard> getAssistantLeaderboard();

	/*
	 * -------------------------------------------------------------------------
	 * Action reporting
	 * -------------------------------------------------------------------------
	 */

	UserActionSummary getUserActionSummary(UUID userId);

	List<UserActionSummary> getUserActionSummaries();

	List<UserWorkflowSummary> getUserWorkflowSummary(UUID userId);

	List<UserWorkflowSummary> getUserWorkflowSummaryByWorkflow(UUID workflowId);

	List<UserWorkflowSummary> getMostUsedWorkflowsByUser(UUID userId);

	List<UserWorkflowSummary> getMostUsedWorkflows();

	WorkflowPopularity getWorkflowPopularity(UUID workflowId);

	List<WorkflowPopularity> getWorkflowPopularity();

	List<WorkflowPopularity> getMostRunWorkflows();

	/*
	 * -------------------------------------------------------------------------
	 * System reporting
	 * -------------------------------------------------------------------------
	 */

	SystemOverview getSystemOverview();

	List<DailySystemStatistics> getDailySystemStatistics(LocalDate from, LocalDate to);

	/*
	 * ------------------------------------------------------------------------- LLM
	 * reporting
	 * -------------------------------------------------------------------------
	 */

	List<LlmDailyMetrics> getLlmDailyMetrics(LocalDate from, LocalDate to);

	List<LlmHourlyMetrics> getLlmHourlyMetrics(LocalDate from, LocalDate to);

	LlmUserMetrics getLlmUserMetrics(UUID userId);

	List<LlmUserMetrics> getLlmUserMetrics();

	LlmAssistantMetrics getLlmAssistantMetrics(UUID assistantId);

	List<LlmAssistantMetrics> getLlmAssistantMetrics();

	LlmSystemOverview getLlmSystemOverview();

}

File: webapp/frontend/app/src/app/analytics/analytics.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { EMPTY, Observable } from 'rxjs';
import { catchError, map } from 'rxjs/operators';

import { AlertService } from '../alert.service';
import { ApiResult } from '../model/api-result';

export interface AnalyticsRow {
	[key: string]: unknown;
}

export interface AnalyticsMetric {
	name: string;
	value: number;
}

export interface AnalyticsDataset {
	name: string;
	rows: AnalyticsRow[];
}

@Injectable({
	providedIn: 'root'
})
export class AnalyticsService {

	private static readonly BaseUrl = 'api/analytics';
	private static readonly GetDailySystemStatistics = 'api/analytics/system/daily';

	constructor(
		private http: HttpClient,
		private alertService: AlertService
	) {
	}

	// -------------------------------------------------------------------------
	// User reporting
	// -------------------------------------------------------------------------

	getUserLoginSummary(userId: string): Observable<AnalyticsRow> {
		return this.get<AnalyticsRow>(
			`${AnalyticsService.BaseUrl}/users/${userId}/login`
		);
	}

	getUserLoginSummaries(): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/users/login`
		);
	}

	getUserActivity(userId: string): Observable<AnalyticsRow> {
		return this.get<AnalyticsRow>(
			`${AnalyticsService.BaseUrl}/users/${userId}/activity`
		);
	}

	getUserActivitySummaries(): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/users/activity`
		);
	}

	getUserAssistantUsage(userId: string): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/users/${userId}/assistants`
		);
	}

	// -------------------------------------------------------------------------
	// User actions
	// -------------------------------------------------------------------------

	getUserActionSummary(userId: string): Observable<AnalyticsRow> {
		return this.get<AnalyticsRow>(
			`${AnalyticsService.BaseUrl}/users/${userId}/actions`
		);
	}

	getUserActionSummaries(): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/users/actions`
		);
	}

	// -------------------------------------------------------------------------
	// Workflow reporting
	// -------------------------------------------------------------------------

	getUserWorkflows(userId: string): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/users/${userId}/workflows`
		);
	}

	getMostUsedWorkflowsByUser(userId: string): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/users/${userId}/workflows/most-used`
		);
	}

	getWorkflowUsers(workflowId: string): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/workflows/${workflowId}/users`
		);
	}

	getWorkflowPopularity(): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/workflows`
		);
	}

	getMostRunWorkflows(): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/workflows/most-run`
		);
	}

	getWorkflowPopularityById(workflowId: string): Observable<AnalyticsRow> {
		return this.get<AnalyticsRow>(
			`${AnalyticsService.BaseUrl}/workflows/${workflowId}`
		);
	}

	// -------------------------------------------------------------------------
	// Assistant reporting
	// -------------------------------------------------------------------------

	getAssistantPopularity(): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/assistants`
		);
	}

	getAssistantPopularityById(
		assistantId: string
	): Observable<AnalyticsRow> {
		return this.get<AnalyticsRow>(
			`${AnalyticsService.BaseUrl}/assistants/${assistantId}`
		);
	}

	getAssistantLeaderboard(): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/assistants/leaderboard`
		);
	}

	getAssistantDailyUsage(
		from: Date,
		to: Date
	): Observable<AnalyticsRow[]> {
		return this.getWithDates<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/assistants/daily`,
			from,
			to
		);
	}

	getAssistantDailyUsageForAssistant(
		assistantId: string,
		from: Date,
		to: Date
	): Observable<AnalyticsRow[]> {
		return this.getWithDates<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/assistants/${assistantId}/daily`,
			from,
			to
		);
	}

	// -------------------------------------------------------------------------
	// System reporting
	// -------------------------------------------------------------------------

	getSystemOverview(): Observable<AnalyticsRow> {
		return this.get<AnalyticsRow>(
			`${AnalyticsService.BaseUrl}/system`
		);
	}

	getDailySystemStatistics(from: Date, to: Date): Observable<AnalyticsRow[]> {
		const params = {
			from: this.formatDate(from),
			to: this.formatDate(to)
		};

		return this.http
			.get<ApiResult>(
				AnalyticsService.GetDailySystemStatistics,
				{ params }
			)
			.pipe(
				catchError(error => {
					this.alertService.postFailure(JSON.stringify(error));
					return EMPTY;
				}),
				map((result: ApiResult) => {
					return (result.data ?? []) as AnalyticsRow[];
				})
			);
	}
	// -------------------------------------------------------------------------
	// LLM reporting
	// -------------------------------------------------------------------------

	getLlmDailyMetrics(
		from: Date,
		to: Date
	): Observable<AnalyticsRow[]> {
		return this.getWithDates<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/llm/daily`,
			from,
			to
		);
	}

	getLlmHourlyMetrics(
		from: Date,
		to: Date
	): Observable<AnalyticsRow[]> {
		return this.getWithDates<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/llm/hourly`,
			from,
			to
		);
	}

	getLlmUserMetrics(): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/llm/users`
		);
	}

	getLlmUserMetricsById(userId: string): Observable<AnalyticsRow> {
		return this.get<AnalyticsRow>(
			`${AnalyticsService.BaseUrl}/llm/users/${userId}`
		);
	}

	getLlmAssistantMetrics(): Observable<AnalyticsRow[]> {
		return this.get<AnalyticsRow[]>(
			`${AnalyticsService.BaseUrl}/llm/assistants`
		);
	}

	getLlmAssistantMetricsById(
		assistantId: string
	): Observable<AnalyticsRow> {
		return this.get<AnalyticsRow>(
			`${AnalyticsService.BaseUrl}/llm/assistants/${assistantId}`
		);
	}

	getLlmSystemOverview(): Observable<AnalyticsRow> {
		return this.get<AnalyticsRow>(
			`${AnalyticsService.BaseUrl}/llm/system`
		);
	}

	// -------------------------------------------------------------------------
	// HTTP helpers
	// -------------------------------------------------------------------------

	private get<T>(url: string): Observable<T> {
		return this.http.get<ApiResult>(url)
			.pipe(
				catchError(error => {
					this.alertService.postFailure(JSON.stringify(error));
					return EMPTY;
				}),
				map(result => result.data as T)
			);
	}

	private getWithDates<T>(
		url: string,
		from: Date,
		to: Date
	): Observable<T> {

		const params = new HttpParams()
			.set('from', this.formatDate(from))
			.set('to', this.formatDate(to));

		return this.http.get<ApiResult>(url, { params })
			.pipe(
				catchError(error => {
					this.alertService.postFailure(JSON.stringify(error));
					return EMPTY;
				}),
				map(result => result.data as T)
			);
	}

	private formatDate(date: Date): string {
		const year = date.getFullYear();
		const month = String(date.getMonth() + 1).padStart(2, '0');
		const day = String(date.getDate()).padStart(2, '0');

		return `${year}-${month}-${day}`;
	}
}